### PR TITLE
feat(iotevents): support comparison operators

### PR DESCRIPTION
**Pull request by yamatatsu**
This PR includes to implement comparison operators.

- [x] integ test

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iotevents/lib/expression.ts [hunk 1]
```diff
@@ -70,6 +112,6 @@ class BinaryOperationExpression extends Expression {
   }
 
   public evaluate() {
-    return `${this.left.evaluate()} ${this.operator} ${this.right.evaluate()}`;
+    return `(${this.left.evaluate()} ${this.operator} ${this.right.evaluate()})`;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/lib/expression.ts:115]: Why this change (to add the parenthesis)?

**Review by skinny85 — DISMISSED** [review]
Looks good @yamatatsu! One question.

File: packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts [hunk 2]
```diff
@@ -40,6 +40,17 @@ class TestStack extends cdk.Stack {
...
+        eventName: 'test-input-event',
+        actions: [
+          ['neq', iotevents.Expression.neq(iotevents.Expression.fromString('31.7'), iotevents.Expression.fromString('31.7'))] as const,
+          ['gt', iotevents.Expression.gt(iotevents.Expression.fromString('31.7'), iotevents.Expression.fromString('31.7'))] as const,
+          ['lt', iotevents.Expression.lt(iotevents.Expression.fromString('31.7'), iotevents.Expression.fromString('31.7'))] as const,
+          ['gte', iotevents.Expression.gte(iotevents.Expression.fromString('31.7'), iotevents.Expression.fromString('31.7'))] as const,
+          ['lte', iotevents.Expression.lte(iotevents.Expression.fromString('31.7'), iotevents.Expression.fromString('31.7'))] as const,
+          ['or', iotevents.Expression.or(iotevents.Expression.fromString('true'), iotevents.Expression.fromString('false'))] as const,
+        ].map(([variableName, value]) => ({ bind: () => ({ configuration: { setVariable: { variableName, value: value.evaluate() } } }) })),
+      }],
```
Comment (yamatatsu) [inline packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts:53]: @skinny85 
I have added integ test but it is efficient?
I wanted to confirm that these expressions would work on AWS actually. And its purpose is filled. But I wonder this integ code have value for aws-cdk.
WDYT?

File: packages/@aws-cdk/aws-iotevents/lib/expression.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-iotevents/lib/expression.ts:115]: Right, but this is not the way to solve this I think 🙂.

I think we should add a `priority` property to all of the expression classes, and each expression will have a different one (`and` will have `2`, `or` will have `1`, etc.). Then, if a sub-expression has a lower-priority then its parent, we will add the parenthesis, if not - we won't.

Hope this makes sense!

**Comment by skinny85** [comment]
@yamatatsu note that I'll be on vacation for the next 2 weeks, so my responses here might be delayed. Thanks for your patience!

File: packages/@aws-cdk/aws-iotevents/lib/expression.ts (hunk 1 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/lib/expression.ts:115]: I agree that the solution (adding priority) can solve the problem.
But I think the first solution I wrote would certainly evaluate them in the order intended by the developer with less code. Because developers don't write `Expression.and(false, Expression.or(true, true))` for they intend to say `(false && true) || true`.

The second solution provide adding less parenthesis automatically like linter auto fixing. But It may be seem as magic. When expressions not work, developers may doubt parenthesis (whichiever the parenthesis caused or not).
The expressions first solution genarated will more descriptive for order of evaluation I think (but verbose?).

Above is my thinking.
Whichever solution we choose, I will spare no effort!

I wish you a good vacation! 😄

Comment (skinny85) [reply packages/@aws-cdk/aws-iotevents/lib/expression.ts:115]: @yamatatsu  Hmm. But don't the extra parenthesis actually inhibit readability?

In your example, `Expression.and(false, Expression.or(true, true))` will be rendered as `(false && (true || true))`. Obviously, we can make it `false && (true || true)` without changing the meaning. And it would be much more pronounced if the expression was more complex (every binary operation would add their own parenthesis).

So I feel like it's the extra parenthesis that would cause the developers to not like these generated expressions 🙂.

And a side benefit? It will not cause any changes to the existing templates, so will be very nice for customers when upgrading the CDK version once this lands 🙂.

Can you at least give it a shot with the `priority` attribute, and see how it feels? If it's very complicated, I'll let this go, and live with the extra parenthesis, but I'd love to at least give it a shot 🙂.

Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/lib/expression.ts:115]: OK! I'll implement it👍🏻

File: packages/@aws-cdk/aws-iotevents/lib/expression.ts [hunk 3]
```diff
@@ -65,11 +114,21 @@ class BinaryOperationExpression extends Expression {
     private readonly left: Expression,
     private readonly operator: string,
     private readonly right: Expression,
+    /**
+     * Indicates the priority of the operator.
+     * This is intended to be set according to MDN rules.
+     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table
```
Comment (yamatatsu) [inline packages/@aws-cdk/aws-iotevents/lib/expression.ts:120]: I have adopted MDN as the operator priority rule. I think it can be used to whatever follow general rules and has concrete numbers.

File: packages/@aws-cdk/aws-iotevents/lib/expression.ts (hunk 1 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/lib/expression.ts]: @skinny85 
I've done!

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks fantastic @yamatatsu! Super clean, and doesn't render the extra parenthesis. Thanks for the extra effort you put into this!

I have a few last stylistic comments, but other than that, we're good to go here!

**Review by skinny85 — APPROVED** [review]
Thanks for the contribution @yamatatsu!

**Review by skinny85 — APPROVED** [review]
